### PR TITLE
feat: filter ckb unlock events

### PR DESCRIPTION
**Pull request by pygman**

### CODE REVIEW

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 1]
```diff
@@ -743,6 +815,75 @@ export class CkbHandler {
...
+    let chain;
+    let burnIds;
+    try {
+      chain = unlockMemo.getXchain();
+      const burnIdBytes: BytesVec = unlockMemo.getBurnIds();
+      burnIds = lodash.range(burnIdBytes.length()).map((i) => new Reader(burnIdBytes.indexAt(i).raw()).serializeJson());
+    } catch (e) {
+      logger.warn(`parse recipient unlock memo data error: ${e.message} ${e.stack}`);
+      return null;
+    }
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:877]: Don't swallow error, which brings bug here

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 2]
```diff
@@ -743,6 +815,75 @@ export class CkbHandler {
...
+
+    const ethereumBurns: EthereumBurn[] = await this.db.getEthereumBurnByUniqueIds(burnIds);
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:879]: Don't get info from db, get it from tx.

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 3]
```diff
@@ -743,6 +815,75 @@ export class CkbHandler {
     }
   }
 
+  async parseUnlockTx(tx: Transaction): Promise<NervosUnlockAssetTxMetaData | null> {
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:818]: The right `parseUnlockTx` logic:

1. verify: tx input[0] should be committeeMultisigCkbCell or committeeMultisigSudtCell
2. verify: witnesses.length === input.length + 1 && can parse unlockMemo from witnesses[input.length] succeed
3. get info: get burnIds from witness
4. get info: get unlock info from outputs[0...burnIds.length], the burnId and outputCell are one-to-one

Besides, the asset of unlockedCells should  be one kind asset(CKB or one kind sUDT)

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 1 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts:877]: fixed, try only parse unlockMemo and log error

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 2 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts:879]: fixed, parse unlock info from transaction

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 3 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts:818]: fixed,
1. inputs[0] === committeeMultisigCkbCell or committeeMultisigSudtCell
2. verify witnesses length, and try parse unlockMemo
3. get burnIds from witness
4. parse output info, verify same kind asset

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 4]
```diff
@@ -1032,6 +1076,166 @@ export class CkbHandler {
...
+    const burnIds = lodash
+      .range(burnIdBytes.length())
+      .map((i) => new Reader(burnIdBytes.indexAt(i).raw()).serializeJson());
+
+    if (burnIds.length > tx.outputs.length - 2) {
+      logger.warn(
+        `invalid unlock tx: burnIds in unlock memo are more than outputs: burnIds: ${burnIds}, outputs: ${tx.outputs}`,
+      );
+      return null;
+    }
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1167]:
```suggestion
    if (burnIds.length >= tx.outputs.length - 2) {
      logger.error(
        `invalid unlock tx: burnIds in unlock memo are more than outputs: burnIds: ${burnIds}, outputs: ${tx.outputs}`,
      );
      return null;
    }
```

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 5]
```diff
@@ -1032,6 +1076,166 @@ export class CkbHandler {
...
+    const senderTypescript = senderOutput.type;
+    if (
+      senderTypescript &&
+      (senderTypescript.codeHash !== sudtTypescript.codeHash || senderTypescript.hashType !== sudtTypescript.hashType)
+    ) {
+      logger.warn(`invalid unlock tx: the typescript of first input is not null or sudt typescript`);
+      return null;
+    }
+
+    for (let i = 1; i < tx.inputs.length; i++) {
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1118]: No need to verify inputs[1..]

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 6]
```diff
@@ -1032,6 +1076,166 @@ export class CkbHandler {
...
+    const assetIdent = isCkb
+      ? '0x0000000000000000000000000000000000000000000000000000000000000000'
+      : utils.computeScriptHash({
+          code_hash: senderTypescript!.codeHash,
+          hash_type: senderTypescript!.hashType,
+          args: senderTypescript!.args,
+        });
+
+    const iCkbUnlocks: ICkbUnlock[] = [];
+    const ethBurns: EthBurn[] = await this.db.getEthBurnByUniqueIds(burnIds);
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1178]: Don't get data from EthBurn, EthBurn record maybe null. We just get unlock_related_data from tx outputs, and save it to db

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 7]
```diff
@@ -6,20 +6,21 @@ import { RPC } from '@ckb-lumos/rpc';
 import { BigNumber } from 'bignumber.js';
 import { Reader } from 'ckb-js-toolkit';
 import * as lodash from 'lodash';
+import { TransferOutSwitch } from '../audit/switch';
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:9]: It's not switch for ckbUnlock, we can ignore switch for now, and add that feature later

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 4 above)
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1167]:
```suggestion
```

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 5 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: removed inputs[1...] check

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 6 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: fixed, get burnTxHash by splite from burnId

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 7 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: removed this import

File: offchain-modules/packages/x/src/handlers/ckb.ts (hunk 4 above)
Comment (pygman) [reply offchain-modules/packages/x/src/handlers/ckb.ts]: removed length check, but check lockscript every output

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 8]
```diff
@@ -1032,6 +1075,139 @@ export class CkbHandler {
...
+          hash_type: senderTypescript!.hashType,
+          args: senderTypescript!.args,
+        });
+
+    const iCkbUnlocks: ICkbUnlock[] = [];
+    for (let i = 0; i < burnIds.length; i++) {
+      const burnId = burnIds[i];
+      const burnHashSeparatorIndex = burnId.indexOf('-');
+      if (burnHashSeparatorIndex < 0) {
+        logger.warn(`invalid unlock tx: invalid burnId in unlockMemoWitness, tx: ${tx}, burnId: ${burnId}`);
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1151]:
```suggestion
        logger.error(`invalid unlock tx: invalid burnId in unlockMemoWitness, tx: ${tx}, burnId: ${burnId}`);
```

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 9]
```diff
@@ -1032,6 +1075,139 @@ export class CkbHandler {
...
+      if (!output) {
+        logger.error(
+          `invalid unlock tx: burnIds in unlock memo are more than outputs: burnIds: ${burnIds}, outputs: ${tx.outputs}, i: ${i}, output: ${output}`,
+        );
+        return null;
+      }
+      let amount: string;
+      if (isCkb) {
+        if (output.type) {
+          logger.warn(`invalid unlock tx: typescript of output should be null when is ckb: ${output}`);
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1165]:
```suggestion
          logger.error(`invalid unlock tx: typescript of output should be null when is ckb: ${output}`);
```

File: offchain-modules/packages/x/src/handlers/ckb.ts [hunk 10]
```diff
@@ -1032,6 +1075,139 @@ export class CkbHandler {
...
+        }
+        amount = output.capacity;
+      } else {
+        if (
+          !output.type ||
+          output.type.codeHash !== senderTypescript?.codeHash ||
+          output.type.hashType !== senderTypescript?.hashType ||
+          output.type.args !== senderTypescript?.args
+        ) {
+          logger.warn(
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/handlers/ckb.ts:1176]:
```suggestion
          logger.error(
```

File: offchain-modules/packages/x/src/db/model.ts [hunk 11]
```diff
@@ -305,6 +312,23 @@ export interface NervosLockAssetTxMetaData {
...
+  amount: bigint;
+  id: string;
+  burnTxHash: string;
+  burnBlockHeight: number;
+}
+
+export interface NervosUnlockAssetTxRecords {
+  txHash: string;
+  records: NervosUnlockAssetTxRecord[];
+}
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/db/model.ts:330]: Useless code ?

Comment (pygman) [reply offchain-modules/packages/x/src/db/model.ts]: fixed

**Pull request merged**